components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt

```diff
@@ -148,16 +150,47 @@ class GeckoEngine(
         onSuccess: ((WebExtension) -> Unit),
         onError: ((String, Throwable) -> Unit)
     ) {
-        GeckoWebExtension(id, url, allowContentMessaging, supportActions).also { ext ->
-            runtime.registerWebExtension(ext.nativeExtension).then({
-                webExtensionDelegate?.onInstalled(ext)
-                onSuccess(ext)
-                GeckoResult<Void>()
-            }, {
-                throwable -> onError(id, throwable)
-                GeckoResult<Void>()
+        val ext = GeckoWebExtension(id, url, allowContentMessaging, supportActions)
+        installWebExtension(ext, onSuccess, onError)
+    }
+
+    internal fun installWebExtension(
+        ext: GeckoWebExtension,
+        onSuccess: ((WebExtension) -> Unit) = { },
+        onError: ((String, Throwable) -> Unit) = { _, _ -> }
+    ) {
+        if (ext.supportActions) {
+            // We currently have to install the global action handler before we
+            // install the extension which is why this is done here directly.
+            // This code can be removed from the engine once the new GV addon
+            // management API lands. Then the global handlers will be invoked
+            // with the latest state whenever they are registered:
+            // https://bugzilla.mozilla.org/show_bug.cgi?id=1599897
+            // https://bugzilla.mozilla.org/show_bug.cgi?id=1582185
+            ext.registerActionHandler(object : ActionHandler {
+                override fun onBrowserAction(extension: WebExtension, session: EngineSession?, action: BrowserAction) {
+                    webExtensionDelegate?.onBrowserActionDefined(extension, action)
+                }
+
+                override fun onToggleBrowserActionPopup(
+                    extension: WebExtension,
+                    action: BrowserAction
+                ): EngineSession? {
```

Comment:
Based on the comment below, if the return type were to be `EngineSession?`:

```return webExtensionDelegate?.onToggleBrowserActionPopup(extension, GeckoEngineSession(runtime), action)```

Unless, there is a design reason why boolean is returned of course :)

Reply:
Yes, good catch, this is easier to read now!